File: src/net/dinkla/raytracer/objects/Disk.java
package net.dinkla.raytracer.objects;

import net.dinkla.raytracer.hits.Hit;
import net.dinkla.raytracer.hits.ShadowHit;
import net.dinkla.raytracer.math.*;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 12.04.2010
 * Time: 19:06:27
 * To change this template use File | Settings | File Templates.
 */
public class Disk extends GeometricObject {

    public Point3DF center;
    public float radius;
    public Normal normal;

    public Disk(final Point3DF center, final float radius, final Normal normal) {
        this.center = center;
        this.radius = radius;
        this.normal = normal;
    }

    @Override
    public boolean hit(final Ray ray, Hit sr) {
        float nom = center.minus(ray.o).dot(normal);
        float denom = ray.d.dot(normal);
        float t = nom / denom;
        if (t <= MathUtils.K_EPSILON) {
            return false;
        }
        Point3DF p = ray.linear(t);
        if (center.distanceSquared(p) < radius * radius)  {
            sr.setT(t);
            sr.setNormal(normal);
//            sr.localHitPoint = p;
            return true;
        } else {
            return false;
        }
    }

    @Override
    public boolean shadowHit(final Ray ray, ShadowHit tmin) {
        float nom = center.minus(ray.o).dot(normal);
        float denom = ray.d.dot(normal);
        float t = nom / denom;
        if (t <= MathUtils.K_EPSILON) {
            return false;
        }
        Point3DF p = ray.linear(t);
        if (center.distanceSquared(p) < radius * radius)  {
            tmin.setT(t);
            return true;
        } else {
            return false;
        }
    }

    public Normal getNormal(Point3DF p) {
        return normal;
    }

    @Override
    public BBox getBoundingBox() {
        // TODO: more exact bounding box of a disk
        Point3DF p = center.minus(radius);
        Point3DF q = center.plus(radius);
        return new BBox(p, q);
    }

}

File: src/main/kotlin/net/dinkla/raytracer/objects/Disk.kt
package net.dinkla.raytracer.objects

import net.dinkla.raytracer.hits.Hit
import net.dinkla.raytracer.hits.ShadowHit
import net.dinkla.raytracer.math.*
import net.dinkla.raytracer.utilities.equals
import net.dinkla.raytracer.interfaces.hash

open class Disk(val center: Point3D, val radius: Double, val normal: Normal) : GeometricObject() {

    init {
        boundingBox = BBox(center - radius, center + radius)
    }

    override fun hit(ray: Ray, sr: Hit): Boolean {
        val nom = (center - ray.origin) dot normal
        val denom = ray.direction dot normal
        val t = nom / denom
        if (t <= MathUtils.K_EPSILON) {
            return false
        }
        val p = ray.linear(t)
        if (center.sqrDistance(p) < radius * radius) {
            sr.t = t
            sr.normal = normal
            //            sr.localHitPoint = p;
            return true
        } else {
            return false
        }
    }

    override fun shadowHit(ray: Ray, tmin: ShadowHit): Boolean {
        val nom = (center - ray.origin) dot normal
        val denom = ray.direction dot normal
        val t = nom / denom
        if (t <= MathUtils.K_EPSILON) {
            return false
        }
        val p = ray.linear(t)
        if (center.sqrDistance(p) < radius * radius) {
            tmin.t = t
            return true
        } else {
            return false
        }
    }

    // TODO why with p?
    fun getNormal(p: Point3D): Normal = normal

    override fun equals(other: Any?): Boolean = this.equals<Disk>(other) { a, b ->
        a.center == b.center && a.radius == b.radius && a.normal == b.normal
    }

    override fun hashCode(): Int = this.hash(center, radius, normal)

    override fun toString(): String = "Disk($center, $radius, $normal)"
}
